fix(cache): Ensure execution cache remains locked until updated

yongkangc opened the pull request:
closes #18541

**Problem:**
- Prevent a possible race condition where the ExecutionCache can be updated while a background prewarming task is still using it by another background prewarming task. 
-  e.g  having a fork block and for `cache_for` to be called before the previous prewarm task exits


**Solution** - using `Arc<()>` as a lightweight, reference-counted lock:
- When the `PrewarmCacheTask` is spawned, it is given its own clone of the `usage_guard`.
- This task holds onto the guard for its entire lifetime. As long as the task is running, the `strong_count` remains > 1, keeping the cache safely locked.
- When the task finishes, its guard is dropped automatically (RAII), decrementing the count and helping to release the lock.

yongkangc commented on crates/engine/tree/src/tree/payload_processor/mod.rs:
```diff
@@ -556,17 +569,21 @@ impl Drop for CacheTaskHandle {
-            .as_ref()
-            .and_then(|cache| (cache.executed_block_hash() == parent_hash).then(|| cache.clone()))
+    /// Returns the idle cache for `parent_hash` with a lock guard, or None if unavailable.
+    pub(crate) fn get_cache_for(&self, parent_hash: B256) -> Option<(SavedCache, Arc<()>)> {
```
hmm this makes more sense, thanks!

yongkangc commented on crates/engine/tree/src/tree/payload_processor/mod.rs:
```diff
@@ -561,12 +561,13 @@ struct ExecutionCache {
-        cache
-            .as_ref()
-            .and_then(|cache| (cache.executed_block_hash() == parent_hash).then(|| cache.clone()))
+        cache.as_ref().and_then(|cache| {
+            (cache.executed_block_hash() == parent_hash && cache.is_available())
+                .then(|| cache.clone())
```
I've refactored it to use filter().cloned() which makes the intent clearer. Thanks for pointing this out!

mattsse commented on crates/engine/tree/src/tree/cached_state.rs:
```diff
@@ -520,16 +520,16 @@ pub(crate) struct SavedCache {
+
+    /// A guard to track in-flight usage of this cache.
+    /// The cache is considered available if the strong count is 1.
+    pub(crate) usage_guard: Arc<()>,
```
can we make this super private, so we don't have to think about where else this can be accessed?

mattsse commented on crates/engine/tree/src/tree/payload_processor/mod.rs:
```diff
@@ -357,15 +357,14 @@ where
-                    ExecutionCacheBuilder::default().build_caches(self.cross_block_cache_size);
-                SavedCache::new(parent_hash, cache, CachedStateMetrics::zeroed())
-            })
+        if let Some(cache) = self.execution_cache.get_cache_for(parent_hash) {
+            debug!("reusing execution cache");
+            cache
+        } else {
+            debug!("creating new execution cache on cache miss");
+            let cache = ExecutionCacheBuilder::default().build_caches(self.cross_block_cache_size);
+            SavedCache::new(parent_hash, cache, CachedStateMetrics::zeroed())
+        }
```
there's no actual change here besides replacing this with let some else?

mattsse reviewed (changes requested):
we need to do a bit more work, because rn, this doesn't take effect because of:

https://github.com/paradigmxyz/reth/blob/573957895fe45924447a9348412b5db508fcf0e1/crates/engine/tree/src/tree/payload_processor/mod.rs#L317-L317

and:

https://github.com/paradigmxyz/reth/blob/573957895fe45924447a9348412b5db508fcf0e1/crates/engine/tree/src/tree/payload_processor/prewarm.rs#L249-L249

so what we should do is

change how we store this is the prewarm context which is held until we finally update the cache at the end:

https://github.com/paradigmxyz/reth/blob/573957895fe45924447a9348412b5db508fcf0e1/crates/engine/tree/src/tree/payload_processor/prewarm.rs#L158-L162

yongkangc commented:
@mattsse thanks for pointing out the flaw in the lifetime.

addressed it [in this commit here](https://github.com/paradigmxyz/reth/pull/18564/commits/65fc90227478a1eec7a1065d6faa11f64476c756) where we moved the usage guard into `PrewarmContext` and only released during `saved_cache`
